src/lang/base/Recursion.ml

```diff
@@ -91,18 +91,35 @@ module NatRec = struct
 
     module Foldr = struct
       (* The type of the fixpoint argument *)
-      let fix_type = parse_type "('A -> 'B -> 'B) -> 'B -> (List 'A) -> 'B"
+      let fix_type = parse_type "('A -> 'B -> 'B) -> ('B -> 'B) -> 'B -> (List 'A) -> 'B"
```

Comment:
I don't think the type should change. When `fold_right` is used from a contract the user would expect it to have the type we had before.

What I had in mind was to define a local function in CPS, something like the following OCaml code:

```
let foldr f z l =
  let helper l k =
    match l with
    | [] -> k z
    | h :: t -> 
      let kk = (fun zz -> k (f h zz)) in
      helper t kk
  in
  helper l (fun x -> x)
```

The function `helper` then performs the actual traversal in CPS, but the continuation parameter is not exposed to the outside world.

Reply:
1. The type to the `list_foldr` that is exposed to the outside world hasn't changed.
2. What you suggest cannot be done directly because we cannot have fixpoints in Scilla, except through constructing them indirectly in OCaml (which is what I've done).

Reply:
You're right, my mistake. Looking at your code again it works pretty much as what I wrote.